### PR TITLE
Preview arbitrary urls within a document

### PR DESCRIPTION
**Pull request by apoorv-mishra**
Enhances #5109 

- [x] Presenter named `presentUnfurl` or something similar
- [x] Iframely [error handling](https://iframely.com/docs/result-codes)
- [x] Integrate with FE
- [x] Tests

Further improvements:

- [x] Iframely API returns back [`cache_age`](https://oembed.com/) field which can be used to serve the cached response from our end rather than making calls to Iframely API

### CODE REVIEW

**Comment by apoorv-mishra** [comment]
Requested review specifically for UI improvements now that the FE integration is done.

**Comment by tommoor** [comment]
Can we try making the image go all the way to the edges? Should also fix this issue where the last line of text is awkwardly chopped off. Also see if the size of the image can be reserved based on the thumbnail dimensions that come back from the API as when it loads in the size of the card changes.

image

File: server/presenters/unfurls/unfurl.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { IframelyErrorResponse, Unfurl } from "@shared/types";
+
+function presentUnfurl(data: any): Unfurl | IframelyErrorResponse {
+  return !data.error
```
Comment (tommoor) [inline server/presenters/unfurls/unfurl.ts:4]: I don't like this overloading with the error response, but it won't be needed with the other suggestion

File: server/routes/api/urls/urls.ts [hunk 2]
```diff
@@ -47,22 +49,24 @@ router.post(
     }
 
     const previewDocumentId = parseDocumentSlug(url);
-    if (!previewDocumentId) {
-      ctx.response.status = 204;
+    if (previewDocumentId) {
+      const document = previewDocumentId
+        ? await Document.findByPk(previewDocumentId)
```
[the remark this answers is not in the document]
Comment (apoorv-mishra) [reply server/routes/api/urls/urls.ts:54]: Uh-oh!

File: app/components/HoverPreview/HoverPreviewLink.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  CARD_WIDTH,
+  THUMBNAIL_HEIGHT,
+} from "./Components";
+
+type Props = {
+  /** Link url */
+  url: string;
+  /** Title for the preview card */
+  title: string;
+  /** Url for thumbnail served by the link provider*/
```
Comment (tommoor) [inline app/components/HoverPreview/HoverPreviewLink.tsx:20]:
```suggestion
  /** Url for thumbnail served by the link provider */
```

File: app/components/HoverPreview/HoverPreviewLink.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  thumbnailUrl: string;
+  /** Some description about the link provider */
+  description: string;
+};
+
+function HoverPreviewLink({ url, thumbnailUrl, title, description }: Props) {
+  return (
+    <Preview to={{ pathname: url }} target="_blank" rel="noopener noreferrer">
```
Comment (tommoor) [inline app/components/HoverPreview/HoverPreviewLink.tsx:28]:
```suggestion
    <Preview as="a" href={url} target="_blank" rel="noopener noreferrer">
```

Not tested, but this should work

File: plugins/iframely/server/iframely.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    const val = await Redis.defaultClient.get(this.cacheKey(url));
+    if (val) {
+      return JSON.parse(val);
+    }
+  }
+
+  public static async get(url: string) {
+    try {
+      const cached = await this.cached(url);
+      return cached ? cached : this.cache(url);
```
Comment (tommoor) [inline plugins/iframely/server/iframely.ts:49]: It's weird that the `fetch` method is inside `cache`, lets just split those apart. JSDoc comments appreciated on the publicly exported interface here

File: server/env.ts [hunk 6]
```diff
@@ -601,6 +601,24 @@ export class Environment {
...
+    protocols: ["http", "https"],
+  })
+  public IFRAMELY_URL = process.env.IFRAMELY_URL ?? "https://iframe.ly";
+
+  /**
+   * Iframely API key
+   */
+  @IsOptional()
+  @CannotUseWithout("IFRAMELY_URL")
+  public IFRAMELY_API_KEY = process.env.IFRAMELY_API_KEY ?? "";
```
Comment (tommoor) [inline server/env.ts:620]:
```suggestion
  public IFRAMELY_API_KEY = this.toOptionalString(process.env.IFRAMELY_API_KEY);
```

This _is_ optional, empty strings lead to bugs 😉

File: server/utils/unfurl.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+glob
+  .sync(path.join(rootDir, "plugins/*/server/unfurl.js"))
+  .forEach((filePath: string) => {
+    // eslint-disable-next-line @typescript-eslint/no-var-requires
+    const resolver: UnfurlResolver = require(path.join(
+      process.cwd(),
+      filePath
+    ));
+    const name = startCase(filePath.split("/")[2]);
+    resolvers[name] = resolver;
```
Comment (tommoor) [inline server/utils/unfurl.ts:20]: I believe we need to filter out resolvers based on `requiredEnvVars` here, it doesn't happen automatically. It is possible that there will be no valid resolvers and that case can then be handled

**Comment by tommoor** [comment]
Looking good, my comments are mainly about handling the case of these environment variables not being set properly as all existing installations will be missing them.

File: app/components/HoverPreview/HoverPreviewLink.tsx (hunk 4 above)
Comment (apoorv-mishra) [reply app/components/HoverPreview/HoverPreviewLink.tsx:28]: Not as it is, but with some other modifications, yes :+1:

**Pull request merged**